pkg/controller/node/cidr_allocator_test.go

```diff
@@ -17,334 +17,329 @@ limitations under the License.
+							Name: "node0",
+						},
+					},
+				},
+				Clientset: fake.NewSimpleClientset(),
+			},
+			clusterCIDR: func() *net.IPNet {
+				_, clusterCIDR, _ := net.ParseCIDR("127.123.234.0/24")
+				return clusterCIDR
+			}(),
+			serviceCIDR: func() *net.IPNet {
+				_, clusterCIDR, _ := net.ParseCIDR("127.123.234.0/26")
+				return clusterCIDR
+			}(),
+			subNetMaskSize:        30,
+			allocatedCIDRs:        []string{"127.123.234.64/30", "127.123.234.68/30", "127.123.234.72/30", "127.123.234.80/30"},
+			expectedAllocatedCIDR: "127.123.234.76/30",
+		},
 	}
-}
 
-func TestRangeAllocator_RandomishAllocation(t *testing.T) {
-	_, clusterCIDR, _ := net.ParseCIDR("127.123.234.0/16")
-	a := NewCIDRRangeAllocator(clusterCIDR, 24)
-
-	// allocate all the CIDRs
-	var err error
-	cidrs := make([]*net.IPNet, 256)
-
-	for i := 0; i < 256; i++ {
-		cidrs[i], err = a.AllocateNext()
-		if err != nil {
-			t.Fatalf("unexpected error: %v", err)
+TestCase:
+	for _, tc := range testCases {
+		allocator := NewCIDRRangeAllocator(tc.fakeNodeHandler, tc.clusterCIDR, tc.serviceCIDR, tc.subNetMaskSize)
+		// this is a bit of white box testing
+		if len(tc.allocatedCIDRs) > 0 {
+			for _, allocated := range tc.allocatedCIDRs {
+				_, cidr, err := net.ParseCIDR(allocated)
+				if err != nil {
+					t.Fatalf("%v: unexpected error when parsing CIDR %v: %v", tc.description, allocated, err)
+				}
+				rangeAllocator, ok := allocator.(*rangeAllocator)
+				if !ok {
+					t.Logf("%v: found non-default implementation of CIDRAllocator, skipping white-box test...", tc.description)
+					continue TestCase
+				}
+				err = rangeAllocator.cidrs.occupy(cidr)
+				if err != nil {
+					t.Fatalf("%v: unexpected error when occupying CIDR %v: %v", tc.description, allocated, err)
+				}
+			}
 		}
-	}
-
-	_, err = a.AllocateNext()
-	if err == nil {
-		t.Fatalf("expected error because of fully-allocated range")
-	}
-	// release them all
-	for i := 0; i < 256; i++ {
-		a.Release(cidrs[i])
-	}
-
-	// allocate the CIDRs again
-	rcidrs := make([]*net.IPNet, 256)
-	for i := 0; i < 256; i++ {
-		rcidrs[i], err = a.AllocateNext()
+		err := allocator.AllocateOrOccupyCIDR(tc.fakeNodeHandler.Existing[0])
 		if err != nil {
-			t.Fatalf("unexpected error: %d, %v", i, err)
+			t.Errorf("%v: unexpected error in AllocateOrOccupyCIDR: %v", tc.description, err)
 		}
-	}
-	_, err = a.AllocateNext()
-	if err == nil {
-		t.Fatalf("expected error because of fully-allocated range")
-	}
-
-	if !reflect.DeepEqual(cidrs, rcidrs) {
-		t.Fatalf("expected re-allocated cidrs are the same collection")
-	}
-}
-
-func TestRangeAllocator_AllocationOccupied(t *testing.T) {
-	_, clusterCIDR, _ := net.ParseCIDR("127.123.234.0/16")
-	a := NewCIDRRangeAllocator(clusterCIDR, 24)
-
-	// allocate all the CIDRs
-	var err error
-	cidrs := make([]*net.IPNet, 256)
-
-	for i := 0; i < 256; i++ {
-		cidrs[i], err = a.AllocateNext()
-		if err != nil {
-			t.Fatalf("unexpected error: %v", err)
+		time.Sleep(time.Second)
```

Comment:
This may trigger flakes if Jenkins will be overutilized (which is almost always the case :P).
That's why I would really prefer doing this synchronization properly.

IIUC, this is supposed to be the synchronization with those go-routines responsible for calling: "updateCIDRAllocation", right?
I don't want to pollute production code, with synchronization used only for test purposes, so I think the easiest way to fix it properly is to do:
wait.Until(/* iterate over fakeNodeHandler ... until our expected cidr is seen */)